Filter by List

TaiWilkin opened the pull request:
## Overview

Enables users to filter down to the individual list level. 

Examples for use of this filtering would be the Accord on Fire and Building Safety in Bangladesh consolidating their two separate accounts into one, with users being able to filter to each separate upload; users being able to filter by the different lists for tiers of supply chains uploaded by brands; certification schemes having separate lists for different levels of certification, e.g. Gold, Silver, Bronze, which users can then filter by.

Connects #983 

## Demo

<img width="994" alt="Screen Shot 2021-03-02 at 11 21 10 AM" src="https://user-images.githubusercontent.com/21046714/109681893-eb936080-7b4b-11eb-8da2-7885a9c73502.png">
<img width="642" alt="Screen Shot 2021-03-02 at 11 35 55 AM" src="https://user-images.githubusercontent.com/21046714/109681899-ecc48d80-7b4b-11eb-8342-f9cffc6cb150.png">
<img width="407" alt="Screen Shot 2021-03-02 at 11 36 03 AM" src="https://user-images.githubusercontent.com/21046714/109681908-ee8e5100-7b4b-11eb-8934-1510cc076594.png">
<img width="430" alt="Screen Shot 2021-03-02 at 11 36 10 AM" src="https://user-images.githubusercontent.com/21046714/109681916-efbf7e00-7b4b-11eb-9de3-7fa816b35631.png">

## Testing Instructions

* Run `vagrant ssh` and `./scripts/server`
* Navigate to [/api/docs/#!/contributor-lists/contributor_lists_list](http://localhost:8081/api/docs/#!/contributor-lists/contributor_lists_list)
* Enter 2 as the contributor ID 
    - [x] You should get a list of contributor lists as ID/name pairs
* Navigate to http://localhost:8081/api/docs/#!/facilities/facilities_list
* Set the list's param to 2
    - [x] You should get a list of facilities from the Summer 2019 Compliance List
* Navigate to http://localhost:6543/
    - [x] You should not see the contributor lists
* Select Service Provider A
    - [x] You should see the contributor lists select
* Select the Summer 2019 Compliance List and search
    - [x] You should get a list of facilities from that list
    - [x] The map should load correctly
* Select Civil Society Organization A
    - [x] The previously selected listed should no longer be selected
    - [x] You should see multiple lists in the dropdown
* Select Summer 2019 Apparel List and Summer 2019 Compliance List and search
    - [x] You should get a list of facilities from both lists
* Refresh the page
    - [x] Both contributors and both lists should remain selected
* Navigate to http://localhost:6543/profile/2 
    - [x] The Facility List names should be links
* Click the link
    - [x] The map should open with the appropriate contributor and list selected

## Checklist

- [x] `fixup!` commits have been squashed
- [ ] CI passes after rebase
- [x] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

lederer commented:
Adjusted the spacing between the contributor and list filters. And did a drive-by tweak to bring the PPE info button closer to its label.

lederer commented:
Doh! Hadn't noticed my editor made those formatting changes. Thanks for fixing, @TaiWilkin

jwalgran commented:
I was able to demo this, but requires some thoughtful interleaving with embedded mode (#1260). There was a merge conflict when I tried to combine this PR with #1260

When we are in embedded mode, we will still want to show filter by list, but not indent the field, since the "parent" field is no longer visible.

jwalgran reviewed (commented):
Clear, well structured implementation that nicely follows the existing conventions of the app.

There is a conflict with the develop branch that needs a careful integration. When in embedded mode we should show the list filter, but not indent it.

I did find a bug where returning from a facility details sidebar "tab" to the search results tab results in the clearing of the search results.

![2021-03-03 15 06 30](https://user-images.githubusercontent.com/17363/109879284-9ed68500-7c32-11eb-8bb5-cd467d094957.gif)

TaiWilkin commented on src/app/src/actions/filterOptions.js:
```diff
@@ -44,6 +51,29 @@ export function fetchContributorOptions() {
+        })}`;
+
+        return apiRequest
+            .get(`${url}${qs}`)
```
It looks like this doesn't work for arrays without [manually concatenating](https://stackoverflow.com/a/51444749) it anyway, unfortunately.

jwalgran replied:
Ah. Thanks for the link. Hopefully I will remember that the next time I try to use `params`.

TaiWilkin commented on src/django/api/views.py:
```diff
@@ -3437,3 +3439,45 @@ def list(self, request):
+            for list
+            in FacilityList.objects.filter(
+                source__contributor__id__in=contributors).order_by('name')
+        ]
```
Updated in 7c1a137. I ended up removing the serializer class entirely, because it wan't necessary or cooperative with the format of the results, but I did switch to using the queryset appropriately.

TaiWilkin commented:
@jwalgran Ready for another look. I've updated this to work more smoothly with the `embed` parameter PR and fixed the missing search results bug (good catch!).

jwalgran commented:
Looking at this now.

jwalgran reviewed (approved):
Excellent fixups. I exercised this with and without `embed=1` and it worked as described.
